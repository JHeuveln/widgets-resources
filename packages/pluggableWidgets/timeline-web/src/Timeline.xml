<?xml version="1.0" encoding="utf-8"?>
<widget id="com.mendix.widget.web.timeline.Timeline" pluginWidget="true" needsEntityContext="true" offlineCapable="true"
        supportedPlatform="Web"
        xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../node_modules/mendix/custom_widget.xsd">
    <name>Timeline</name>
    <description>Shows vertical timeline with events</description>
    <icon/>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="General">
                <property key="renderMode" type="enumeration" defaultValue="basic">
                    <caption>Render</caption>
                    <description>Basic: creates the Timeline based on the configured items. Custom: free to model timeline.</description>
                    <enumerationValues>
                        <enumerationValue key="basic">Basic</enumerationValue>
                        <enumerationValue key="custom">Custom</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="data" type="datasource" isList="true">
                    <caption>Data Source</caption>
                    <description/>
                </property>
                <property key="eventTime" type="attribute" dataSource="data">
                    <caption>Event date time attribute</caption>
                    <description>Will be used for grouping by date.</description>
                    <attributeTypes>
                        <attributeType name="DateTime"/>
                    </attributeTypes>
                </property>
                <property key="showDayDivider" type="boolean" defaultValue="false">
                    <caption>Show day divider</caption>
                    <description/>
                </property>
            </propertyGroup>
            <propertyGroup caption="Event">
                <property key="title" type="attribute" dataSource="data" required="false">
                    <caption>Title</caption>
                    <description/>
                    <attributeTypes>
                        <attributeType name="String"/>
                    </attributeTypes>
                </property>
                <property key="description" type="attribute" dataSource="data" required="false">
                    <caption>Description</caption>
                    <description/>
                    <attributeTypes>
                        <attributeType name="String"/>
                    </attributeTypes>
                </property>
                <property key="icon" type="icon" required="false">
                    <caption>Icon</caption>
                    <description>Will be circle if its not selected.</description>
                </property>
            </propertyGroup>
            <propertyGroup caption="Custom">
                <property key="customIcon" type="widgets" dataSource="data" required="false">
                    <caption>Icon</caption>
                    <description>Content of the icon</description>
                </property>
                <property key="customDayDivider" type="widgets" dataSource="data" required="false">
                    <caption>Day Divider</caption>
                    <description>Content of the day divider</description>
                </property>
                <property key="customTitle" type="widgets" dataSource="data" required="false">
                    <caption>Title</caption>
                    <description>Content of the title</description>
                </property>
                <property key="customEventDateTime" type="widgets" dataSource="data" required="false">
                    <caption>Event Time</caption>
                    <description>Content of the event time</description>
                </property>
                <property key="customDescription" type="widgets" dataSource="data" required="false">
                    <caption>Content</caption>
                    <description>Content of the description</description>
                </property>
            </propertyGroup>
        </propertyGroup>
    </properties>
</widget>
