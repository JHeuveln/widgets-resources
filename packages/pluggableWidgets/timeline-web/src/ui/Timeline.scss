$primary-color: #264ae5;
$border-color: #dadcde;
$title-color: #0a1325;
$description-color: #6c717c;
$spacing-large: 30px;
$gutter-size: 15px;
$icon-size: 18px;
$icon-position-left: -($icon-size / 2);

.timeline-date {
    .timeline-date-header {
        display: inline-block;
        width: 110px;
        padding: 8px;
        text-align: center;
        border: 1px solid $border-color;
        border-radius: 30px;
    }
    .timeline-events {
        margin-bottom: 0;
        margin-left: 55px;
        padding: 35px 0;
        border-left: 1px solid $border-color;
        ul {
            padding: 0;
            list-style: none;
            margin-bottom: 0;
        }
        .timeline-event {
            position: relative;
            padding-left: ($gutter-size * 2);
            padding-bottom: 20px;

            .icon-wrapper {
                position: absolute;
                top: 3px;
                left: $icon-position-left;
                display: flex;
                justify-content: center;
                align-items: center;
                width: $icon-size;
                height: $icon-size;

                .glyphicon {
                    font-size: $icon-size;
                }
                img {
                    max-width: $icon-size;
                    max-height: $icon-size;
                }
            }

            .timeline-icon-circle {
                width: $icon-size;
                height: $icon-size;
                border-radius: 50%;
                background-color: $primary-color;
            }

            .content-wrapper {
                .date-time-wrapper {
                    margin-right: 15px;
                    p {
                    }
                }
                .info-wrapper {
                    flex-direction: column;

                    .title {
                        line-height: 20px;
                        color: $title-color;
                    }
                    .description {
                        line-height: 20px;
                        color: $description-color;
                    }
                }
            }
        }
    }

    .no-divider {
        padding-bottom: 0;
        padding-top: 0;
    }
}

.flex-container {
    display: flex;
    overflow: hidden;
    flex: 1;
    flex-direction: row;
}
