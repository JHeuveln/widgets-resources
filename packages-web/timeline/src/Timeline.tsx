import { Component, ReactNode, createElement } from "react";
import { HelloWorldSample } from "./components/HelloWorldSample";

import { TimelineContainerProps } from "../typings/TimelineProps";

import "./ui/Timeline.css";

export default class Timeline extends Component<TimelineContainerProps> {
    render(): ReactNode {
        return <HelloWorldSample sampleText={this.props.sampleText ? this.props.sampleText : "World"} />;
    }
}
