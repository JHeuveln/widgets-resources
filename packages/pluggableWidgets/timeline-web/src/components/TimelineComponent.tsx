import { RenderModeEnum } from "../../typings/TimelineProps";
import { Children, createElement, ReactElement, ReactNode } from "react";
import classNames from "classnames";
import { BasicItemType, CustomItemType, ItemType } from "../Timeline";
import { Icon } from "mendix/components/web/Icon";

export interface TimelineComponentProps {
    data: Map<string | ReactNode, ItemType[]>;
    renderMode: RenderModeEnum;
    showDayDivider: boolean;
}

export default function TimelineComponent(props: TimelineComponentProps): ReactElement {
    return <div className="timeline-wrapper">{getItems(props.data, props.renderMode, props.showDayDivider)}</div>;
}

function getBasicEventsFromDay(eventsOfDay: BasicItemType[]): ReactNode[] {
    return eventsOfDay.map((event, index) => (
        <li className="timeline-event" key={index}>
            <div className="icon-wrapper">
                {event.icon?.value ? <Icon icon={event.icon.value} /> : <div className="timeline-icon-circle" />}
            </div>
            <div className="flex-container content-wrapper">
                <div className="date-time-wrapper">
                    <p>{event.time}</p>
                </div>
                <div className="flex-container info-wrapper">
                    <p className="title">{event.title}</p>
                    <p className="description">{event.description}</p>
                </div>
            </div>
        </li>
    ));
}

function getCustomEventsFromDay(eventsOfDay: CustomItemType[]): { eventDayDivider: ReactNode; events: ReactNode[] } {
    let eventDayDivider: ReactNode = null;
    const events = eventsOfDay.map((event, index) => {
        eventDayDivider = event.dayDivider;
        return (
            <li className="timeline-event" key={index}>
                <div className="icon-wrapper">{event.icon ?? <div className="timeline-icon-circle" />}</div>
                <div className="flexcontainer content-wrapper">
                    {Children.count((event.eventDateTime as ReactElement).props.children) > 0 && (
                        <div className="date-time-wrapper">{event.eventDateTime}</div>
                    )}
                    <div className="flexcontainer info-wrapper">
                        {event.title}
                        {event.description}
                    </div>
                </div>
            </li>
        );
    });
    return { eventDayDivider, events };
}

export function getItems(
    structuredEvents: Map<string | ReactNode, ItemType[]>,
    renderMode: RenderModeEnum,
    showDayDivider: boolean
): ReactNode[] {
    const days: ReactNode[] = [];
    structuredEvents.forEach((eventsOfDay: ItemType[], day: string) => {
        let events;
        let dayDivider;
        if (renderMode === "basic") {
            events = getBasicEventsFromDay(eventsOfDay as BasicItemType[]);
            dayDivider = <span>{day}</span>;
        } else {
            events = getCustomEventsFromDay(eventsOfDay).events;
            dayDivider = getCustomEventsFromDay(eventsOfDay).eventDayDivider;
        }

        const constructedDiv = (
            <div className="timeline-date" key={day}>
                {showDayDivider && Children.count((dayDivider as ReactElement).props.children) > 0 && (
                    <div className="timeline-date-header">{dayDivider}</div>
                )}
                <div className={classNames("timeline-events", !showDayDivider ? "no-divider" : undefined)}>
                    <ul>{events}</ul>
                </div>
            </div>
        );
        days.push(constructedDiv);
    });
    return days;
}
