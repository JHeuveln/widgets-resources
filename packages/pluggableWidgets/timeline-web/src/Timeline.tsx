import { createElement, ReactElement, ReactNode, useMemo } from "react";
import { TimelineContainerProps } from "../typings/TimelineProps";
import "./ui/Timeline.scss";
import { DynamicValue, WebIcon } from "mendix";
import TimelineComponent from "./components/TimelineComponent";

export interface BasicItemType {
    icon?: DynamicValue<WebIcon>;
    title?: string;
    date?: string;
    time?: string;
    description?: string;
}

export interface CustomItemType {
    icon?: ReactNode;
    dayDivider?: ReactNode;
    title?: ReactNode;
    eventDateTime?: ReactNode;
    description?: ReactNode;
}

export type ItemType = BasicItemType | CustomItemType;

export default function Timeline(props: TimelineContainerProps): ReactElement {
    const structuredEvents = useMemo((): Map<string | ReactNode, ItemType[]> => {
        const eventsMap = new Map<string | ReactNode, ItemType[]>();

        props.data.items?.forEach(item => {
            let constructedItem: ItemType;
            const eventTime = props.eventTime(item);
            const date = eventTime.value;
            if (!date) {
                return;
            }
            let dateFormatter = eventTime.formatter;
            let timeFormatter = eventTime.formatter;

            if (dateFormatter?.type === "datetime") {
                dateFormatter = dateFormatter.withConfig({ type: "date" });
                timeFormatter = dateFormatter.withConfig({ type: "time" });
            }

            if (props.renderMode === "basic") {
                constructedItem = {
                    icon: props.icon,
                    title: props.title?.(item).displayValue,
                    date: dateFormatter?.format(eventTime.value),
                    time: timeFormatter?.format(eventTime.value),
                    description: props.description?.(item).displayValue
                };
            } else {
                constructedItem = {
                    icon: props.customIcon?.(item),
                    dayDivider: props.customDayDivider?.(item),
                    title: props.customTitle?.(item),
                    eventDateTime: props.customEventDateTime?.(item),
                    description: props.customDescription?.(item)
                };
            }

            const currentDates = eventsMap.get(dateFormatter?.format(eventTime.value));
            if (currentDates) {
                currentDates.push(constructedItem);
            } else {
                eventsMap.set(dateFormatter?.format(eventTime.value), [constructedItem]);
            }
        });

        return eventsMap;
    }, [props.data]);

    return (
        <TimelineComponent
            data={structuredEvents}
            showDayDivider={props.showDayDivider}
            renderMode={props.renderMode}
        />
    );
}
