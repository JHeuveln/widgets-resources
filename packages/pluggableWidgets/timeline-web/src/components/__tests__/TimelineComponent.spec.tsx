import { shallow } from "enzyme";
import { createElement, ReactNode } from "react";
import { dynamicValue } from "@widgets-resources/piw-utils";
import TimelineComponent, { TimelineComponentProps } from "../TimelineComponent";
import { BasicItemType, CustomItemType } from "../../Timeline";
import { Icon as WebIcon } from "../../__tests__/__mocks__/mendix/components/web/Icon";

jest.mock("mendix/components/web/Icon", () =>
    jest.requireActual("../../__tests__/__mocks__/mendix/components/web/Icon")
);

describe("Timeline", () => {
    const basicData = new Map<string, BasicItemType[]>();
    const customData = new Map<ReactNode, CustomItemType[]>();

    const basicItem: BasicItemType = {
        description: "Basic description",
        time: new Date(1453, 4, 29).toDateString(),
        title: "Basic title"
    };

    const basicItemWithIcon: BasicItemType = {
        ...basicItem,
        // @ts-ignore
        icon: dynamicValue<typeof WebIcon>(WebIcon, false)
    };

    const customItem: CustomItemType = {
        dayDivider: <p>Day Divider</p>,
        title: <p>Title</p>,
        eventDateTime: <p>Date Time</p>,
        description: <p>Description</p>
    };

    const customItemWithIcon: CustomItemType = {
        ...customItem,
        icon: <img src="customIcon" />
    };

    basicData.set(new Date(1453, 4, 29).toDateString(), [basicItem, basicItemWithIcon]);
    customData.set(new Date(1453, 4, 29).toDateString(), [customItem, customItemWithIcon]);

    const basicRenderProps: TimelineComponentProps = {
        data: basicData,
        renderMode: "basic",
        showDayDivider: true
    };
    const customRenderProps: TimelineComponentProps = {
        data: customData,
        renderMode: "custom",
        showDayDivider: true
    };

    it("renders timeline with basic configuration", () => {
        const component = shallow(<TimelineComponent {...basicRenderProps} />);
        expect(component).toMatchSnapshot();
    });
    it("renders timeline with custom configuration ", () => {
        const component = shallow(<TimelineComponent {...customRenderProps} />);
        expect(component).toMatchSnapshot();
    });
    it("hides the timeline header", () => {
        const component = shallow(<TimelineComponent {...basicRenderProps} showDayDivider={false} />);
        expect(component).toMatchSnapshot();
    });
});
