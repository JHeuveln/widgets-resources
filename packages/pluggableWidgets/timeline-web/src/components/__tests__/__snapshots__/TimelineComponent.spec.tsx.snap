// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline hides the timeline header 1`] = `
<div
  className="timeline-wrapper"
>
  <div
    className="timeline-date"
    key="Sun May 29 1453"
  >
    <div
      className="timeline-events no-divider"
    >
      <ul>
        <li
          className="timeline-event"
          key="0"
        >
          <div
            className="icon-wrapper"
          >
            <div
              className="timeline-icon-circle"
            />
          </div>
          <div
            className="flex-container content-wrapper"
          >
            <div
              className="date-time-wrapper"
            >
              <p>
                Sun May 29 1453
              </p>
            </div>
            <div
              className="flex-container info-wrapper"
            >
              <p
                className="title"
              >
                Basic title
              </p>
              <p
                className="description"
              >
                Basic description
              </p>
            </div>
          </div>
        </li>
        <li
          className="timeline-event"
          key="1"
        >
          <div
            className="icon-wrapper"
          >
            <Component
              icon={[Function]}
            />
          </div>
          <div
            className="flex-container content-wrapper"
          >
            <div
              className="date-time-wrapper"
            >
              <p>
                Sun May 29 1453
              </p>
            </div>
            <div
              className="flex-container info-wrapper"
            >
              <p
                className="title"
              >
                Basic title
              </p>
              <p
                className="description"
              >
                Basic description
              </p>
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Timeline renders timeline with basic configuration 1`] = `
<div
  className="timeline-wrapper"
>
  <div
    className="timeline-date"
    key="Sun May 29 1453"
  >
    <div
      className="timeline-date-header"
    >
      <span>
        Sun May 29 1453
      </span>
    </div>
    <div
      className="timeline-events"
    >
      <ul>
        <li
          className="timeline-event"
          key="0"
        >
          <div
            className="icon-wrapper"
          >
            <div
              className="timeline-icon-circle"
            />
          </div>
          <div
            className="flex-container content-wrapper"
          >
            <div
              className="date-time-wrapper"
            >
              <p>
                Sun May 29 1453
              </p>
            </div>
            <div
              className="flex-container info-wrapper"
            >
              <p
                className="title"
              >
                Basic title
              </p>
              <p
                className="description"
              >
                Basic description
              </p>
            </div>
          </div>
        </li>
        <li
          className="timeline-event"
          key="1"
        >
          <div
            className="icon-wrapper"
          >
            <Component
              icon={[Function]}
            />
          </div>
          <div
            className="flex-container content-wrapper"
          >
            <div
              className="date-time-wrapper"
            >
              <p>
                Sun May 29 1453
              </p>
            </div>
            <div
              className="flex-container info-wrapper"
            >
              <p
                className="title"
              >
                Basic title
              </p>
              <p
                className="description"
              >
                Basic description
              </p>
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Timeline renders timeline with custom configuration  1`] = `
<div
  className="timeline-wrapper"
>
  <div
    className="timeline-date"
    key="Sun May 29 1453"
  >
    <div
      className="timeline-date-header"
    >
      <p>
        Day Divider
      </p>
    </div>
    <div
      className="timeline-events"
    >
      <ul>
        <li
          className="timeline-event"
          key="0"
        >
          <div
            className="icon-wrapper"
          >
            <div
              className="timeline-icon-circle"
            />
          </div>
          <div
            className="flexcontainer content-wrapper"
          >
            <div
              className="date-time-wrapper"
            >
              <p>
                Date Time
              </p>
            </div>
            <div
              className="flexcontainer info-wrapper"
            >
              <p>
                Title
              </p>
              <p>
                Description
              </p>
            </div>
          </div>
        </li>
        <li
          className="timeline-event"
          key="1"
        >
          <div
            className="icon-wrapper"
          >
            <img
              src="customIcon"
            />
          </div>
          <div
            className="flexcontainer content-wrapper"
          >
            <div
              className="date-time-wrapper"
            >
              <p>
                Date Time
              </p>
            </div>
            <div
              className="flexcontainer info-wrapper"
            >
              <p>
                Title
              </p>
              <p>
                Description
              </p>
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
